cpp/src/arrow/util/future.h

```diff
@@ -556,6 +559,38 @@ inline bool WaitForAll(const std::vector<Future<T>*>& futures,
   return waiter->Wait(seconds);
 }
 
+/// \brief Create a Future which completes when all of `futures` complete.
+///
+/// The future's result is a vector of the results of `futures`.
+/// Note that this future will never be marked "failed"; failed results
+/// will be stored in the result vector alongside successful results.
+template <typename T>
+Future<std::vector<Result<T>>> All(std::vector<Future<T>> futures) {
+  struct State {
+    explicit State(std::vector<Future<T>> f)
+        : futures(std::move(f)), n_remaining(futures.size()) {}
+
+    std::vector<Future<T>> futures;
+    std::atomic<size_t> n_remaining;
+  };
```

Comment:
All versions of `FutureWaiter` wait (put the thread in a non-runnable state waiting on the CV) if I read it correctly.  This version adds the "after-wait" as a continuation that runs as a callback  when the last future being waited on completes.
![Untitled presentation](https://user-images.githubusercontent.com/1696093/105902075-3e648000-5fc2-11eb-80c0-d59e65d78b23.png)

Reply:
Yes, I meant that perhaps it would be easy to add a non-blocking facility to `FutureWaiter` internals. It is designed to receive "done" notifications from futures after all (though perhaps it's not more efficient than `AddCallback`?).

Reply:
Ok.  Yes, it could be done with the waiters if the waiter were given a "finished future" and it marked it complete when the futures all finished.  Then a WaitForAllAsync could be added.  I'm not sure there would be any advantage though.  The waiter is basically a named callback from the future's perspective.  I'm happy to proceed with this if you think it'll be cleaner.

Reply:
I was just wondering if perhaps `FutureWaiter` would allow a more efficient implementation. If it doesn't, then don't bother.